prow/plugins/mergecommitblocker/mergecommitblocker.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package mergecommitblocker adds a do-not-merge label to pull requests which contain merge commits
+// Merge commits are defined as commits that contain more than one parent commit SHA
+
+package mergecommitblocker
+
+import (
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/labels"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const pluginName = "mergecommitblocker"
+
+// init registers out plugin as a pull request handler
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+// helpProvider provides information on the plugin
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	// Only the Description field is specified because this plugin is not triggered with commands and is not configurable.
+	return &pluginhelp.PluginHelp{
+		Description: fmt.Sprintf(`The merge commit blocker plugin adds the do-not-merge/contains-merge-commit label to pull requests that contain merge commits`),
+	}, nil
+}
+
+func handlePullRequest(pc plugins.Agent, pe github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.Logger, pe)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	AddLabel(owner, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	GetFile(org, repo, filepath, commit string) ([]byte, error)
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+	ListPRCommits(org, repo string, number int) ([]github.RepositoryCommit, error)
+}
+
+// handlePR takes a github client, a pull request event and applies, or removes applicable labels
+func handlePR(gc githubClient, le *logrus.Entry, pe github.PullRequestEvent) error {
+
+	if !isPRChanged(pe) {
+		return nil
+	}
+	// Store all info about the owner, repo, and num of pull request
+	var (
+		owner = pe.PullRequest.Base.Repo.Owner.Login
+		repo  = pe.PullRequest.Base.Repo.Name
+		num   = pe.PullRequest.Number
+	)
+
+	// Use github client to get the commits in the pull request
+	commits, err := gc.ListPRCommits(owner, repo, num)
```

Comment:
I'm worried this will eat up tokens in a major way -- can we use the `git` client instead?

Reply:
Should be possible,
Ran this git command gave me commit hashes, when there were merge commits two hashes were present
```git --no-pager log master...$(git rev-parse --abbrev-ref HEAD) --pretty=%P``
Should be able to use ```git``` client to replicate this ✨

Reply:
You should be able to use

```
git log --min-parents=2 $( git merge-base <pr-merged-into-branch-ref> <target-branch-ref> )..HEAD
```

Or similar

Reply:
hello there not dead, so I would alter the git client to do a git log? Thanks in advance 🙏

Reply:
The `git` client right now takes actions but doesn't try to create `commit` objects or anything, so we could keep the amount of code necessary small by creating a method like:

```go
func (r *Repo) MergeCommitsExistBetween(target, head string) (bool, error) {
```

Where that function would do the equivalent of the command I pasted above and just returns whether or not a merge commit was found.